### ISSUE TITLE
Add resolvers for other test libraries

### ISSUE THREAD

**Issue by bluzi** [bluzi/travis-buddy#3]
Add resolvers and test repositories for other test libraries, such as:
- Karma
- Jest
- Jasmine
- ...

Also, we should add non-js libraries (Travis also supports Python, Ruby and more)

**Comment by mzfr** [comment]
@bluzi Does TravisBuddy support Python? We're looking for way to post comments on GitHub pull requests but we don't really have any testing framework in place.

Can we somehow configure TravisBuddy to post part of the output of Travis' Build Log as a comment.

So for eg. every line in the log that starts with `TravisBuddy: ` could be posted as a comment.

This will generalize the tool to work without any testing frameworks etc.

I'm willing to work on this if the feature doesn't already exist and sounds right to you. :smile: 

P.S: For some context, please see: https://github.com/xbmc/addon-check/pull/34

/cc @razzeee

**Comment by bluzi** [comment]
Hey @mzfr,

Thanks for reaching out :) This is exactly how TravisBuddy behaves, except that it only takes the "interesting" lines from the build log, which are the lines where your test script runs. TravisBuddy dumps the part of the log where TravisCI sets up the environment.

TravisBuddy should support Python (and any language), however as far as I know it never been tested in production, so I suggest you'll just give it a try and see if the comments comes out right, I'll be happy to assist if it won't work as expected. :)

**Comment by Razzeee** [comment]
Will try it this evening when I'm home. Thanks!

**Comment by bluzi** [comment]
@Razzeee No problem, please let me know how it went if you can

**Comment by mzfr** [comment]
I tried it! (see output comment [here](https://github.com/mzfr/Python-Scripts/pull/4#issuecomment-372399038))

It can output what is printed by the python process onto stdout, but @bluzi is there a way to configure the comment template? to remove things like `1st Build` etc. 

Maybe there could be a raw mode, which comments EXACTLY what was outputted by the process? then we could just send out a properly formatted markdown text and TravisBuddy could comment that. (just thinking out loud here)

**Comment by bluzi** [comment]
Good, it works just as I expected. 

I thought about adding templates, and the codebase is ready for that. The idea was to enable template editing via the dashboard at https://www.travisbuddy.com, however, since the dashboard is not implemented yet, there's still no way of achieving that. 

But no worries, enabling just raw mode is as easy as writing few lines of code, and I can definitely do that. Expect it by the next 48 hours, I'll let you know how you should modify `travis.yml` in order to enable raw mode.

Please feel free to share your thoughts and ideas, I feel like this project has came to a phase where it really just need users feedback in order to improve and be perfect, and as you can see, I have time and motivation to implement these ideas.

@mzfr

**Comment by mzfr** [comment]
@bluzi Damn, you're fast! :small_airplane: 

So if we have raw mode, I don't see what else we could want. We can just comment whatever we want to without having to setup our infrastructure or anything.

@Razzeee Do you foresee any issues with this approac?

**Comment by Razzeee** [comment]
Only thing that comes to mind, is that we have nothing like we we did for the labels with the old approach.

Basically labeling the pr either as 'Travis successful' or 'Travis failed'

**Comment by Razzeee** [comment]
@mzfr While it will work, I'm not so happy about it. As us printing markdown into the log in travis build will enable this to work, but the log on travis-ci itself will probably be not as pretty. Not really a blocker but not ideal either.

**Comment by mzfr** [comment]
Oh, right! totally forgot about the labels.

Maybe we can convince @bluzi to add this feature as well :wink: 

We want pull requests to be labeled `Tests Passed` / `Tests Failed` depending on what the current status is, so that the maintainers can just focus on those PRs.

**Comment by mzfr** [comment]
@Razzeee 

> printing markdown into the log in travis build will enable this to work, but the log on travis-ci itself will probably be not as pretty.

That is where templates could come in? 

We just output standard colored lines onto the Travis log and then have TravisBuddy pick those up and wrap them in a template (perhaps from a file `.travis-buddy-template`) and then comment that?

**Comment by Razzeee** [comment]
That should work, as long as that template is flexible, whatever that means in this context. But wasn't clear that this is in scope, as it was only mentioned by @bluzi in context of the dashboard website.

**Comment by mzfr** [comment]
@bluzi Thanks for your prompt response. 

I'm now testing the raw template, and will let you know how it goes.

@Razzeee Do we have any issues in adding @TravisBuddy as a contributor to Kodi's repos?

**Comment by Razzeee** [comment]
I don't think that will be a problem but @martijnKaijser will have the last word

**Comment by mzfr** [comment]
@bluzi That worked! :tada: [See Here](https://github.com/mzfr/Python-Scripts/pull/4#issuecomment-372424290)

Though I think there's some issue with formatting of first line. Compare the comment with the first line of [Travis log](https://travis-ci.org/mzfr/Python-Scripts/builds/352490202)

BTW, I don't think there's any way to get colored lines in comments except for in the diff format. Maybe we could abuse unicode characters (🔴, 🔵 etc.) and use them as bullets? but that'd be something we add in the template file.

**Comment by Razzeee** [comment]
I don't really care about colors in github comments, just to be upfront :)

What's a problem is, that the header
```
Travis tests have failed

Hey Mehtab Zafar,
Please read the following log in order to understand the failure reason.
It'll be awesome if you fix what's wrong and commit the changes.
```

is missing in rawmode and thus we're not pinging and are not nice anymore :/
Not sure if we can just pull that from travis variables, at least I can't find a env variable that does that.

**Comment by bluzi** [comment]
@mzfr 
> Though I think there's some issue with formatting of first line. Compare the comment with the first line of Travis log

Yup, looks like i forgot to trim the output. Will fix it soon. 

@Razzeee @mzfr 
> What's a problem is, that the header...
I thought the whole idea of raw mode is that you're not getting the header, so you can write so you can write your own markdown using the build log.

**Comment by Razzeee** [comment]
The header is fine. I think the main reason is that we want to trim some of
the lines our log returns before it gets relevant. And also not a big fan
of hiding it in a folding box. Correct me please @mzfr

**Comment by bluzi** [comment]
Alright, so just confirm that's really what you want and I'll make the changes. @mzfr

**Comment by mzfr** [comment]
We want the final comment to be something like:

```markdown
## Tests have failed

Hey {@username/full name}, 

We've found the following issues with this code:

{Output from our testing script goes here}
```

I feel the best way to fix this is via a template. TravisBuddy will do some pre-processing of the template to fill in variables like `{pull_creator}` etc. and will return the final form.

What do you guys think?

**Comment by bluzi** [comment]
I can easily replace the raw mode with this template you're suggesting, just be sure this is what you want and I'll set it up.

**Comment by mzfr** [comment]
@bluzi Are you talking about hardcoding exactly this template? or keeping it general? (say from a `.travis-buddy-template` file) Ideally it'd be better that we have a general method in case we need to change templates etc.

**Comment by bluzi** [comment]
Well, at first I thought about hardcoding it for now, but you're right, I'll make a general implementation and will let you know once it's ready (probably in the next few hours)

Then you'll be able to control your template and modify it whenever you want without my help.

**Comment by bluzi** [comment]
@Razzeee @mzfr 

I made the changes as we discussed, you may now use your own template by adding a file named `travis-buddy-failure-template.md` to the root of repository.
If TravisBuddy will find this file in the branch it's commenting on, it'll be used as the comment's template.

The content of the file must be a valid [mustache](https://github.com/janl/mustache.js/) template.
If you don't know the syntax, visit [the mustache documentation](https://mustache.github.io/mustache.5.html) or refer to this [sample template](https://raw.githubusercontent.com/bluzi/name-db/bluzi-travis-buddy-test/travis-buddy-failure-template.md).

Let me know how it went or if you have any questions.

**Comment by Razzeee** [comment]
Something is wrong with the console result, it seems to swallow stuff.
`Addon i icon exists
icoources/images/svg/chargers.svg`

https://github.com/Razzeee/repo-plugins/pull/1#issuecomment-373903664

https://travis-ci.org/Razzeee/repo-plugins/builds/354653785?utm_source=github_status&utm_medium=notification

**Comment by Razzeee** [comment]
**{{author}}** Also seems to not be the github name of a user

**Comment by bluzi** [comment]
Hey @Razzeee,

I don't see what's wrong in the log you attached, can you be more specific? 
Also, as I mentioned before, TravisBuddy only takes the relevant data from the build log, and omitted the parts where TravisCI configures the environment, so only the output of the test commands you're running will be commented. 
If that's the issue - that's expected behavior, but I can make a flag that tell TravisBuddy to include the entire log in the comments.

**Comment by Razzeee** [comment]
Heres an example, travisBuddy prints this in the comment:
![grafik](https://user-images.githubusercontent.com/5943908/37558903-f1d90848-2a1b-11e8-9d2b-30f581d38dbd.png)

![grafik](https://user-images.githubusercontent.com/5943908/37558916-2ea157b2-2a1c-11e8-9d77-8e43eae07077.png)

MIght be due to the raw log having some markup, but interestingly the first rows that get mangled don't have such.

![grafik](https://user-images.githubusercontent.com/5943908/37558934-60e8a7f2-2a1c-11e8-8410-8596e95af5e7.png)

Yes, we want to use @mention if possible

**Comment by bluzi** [comment]
Yup, looks like it's happening because of the markup. I guess we should be able to strip the markup somehow, but it requires some research and probably will take some time. 

As to the @mention - I'll take care of it soon and will give you the name of the variable that will always return a GitHub username.

**Comment by mzfr** [comment]
hey @bluzi  can you please have a look at my [repo](https://github.com/mzfr/test-travis-buddy/tree/tests) I'm unable to get Travis buddy to work. What am I missing ?

**Comment by bluzi** [comment]
Hey,

It's definitely a problem with the new templating feature, i'm getting an error in `mustache` (`Unclosed tag at 249`), it could be a problem with your template (probably not), and it could be something else. 

I'll check later/tomorrow when I have more time and let you know.

**Comment by Razzeee** [comment]
@mzfr your last tag is missing a closing bracket it's `{{/jobs}` but should be `{{/jobs}}`

**Comment by mzfr** [comment]
@razzeee thank you for pointing that out

**Comment by bluzi** [comment]
Nice catch @Razzeee

**Comment by bluzi** [comment]
Hey guys,

Just an update: @TravisBuddy  now supports @mention, by replacing `{{author}}` in `{{pullRequestAuthor}}` (which results in the GitHub username of the user created the PR)
You can see a sample here: https://github.com/bluzi/name-db/pull/501#issuecomment-375912258 (just a result of a test)

I also updated the default template to @mention the author :) 

As to the markup issue, I tried looking at https://github.com/Razzeee/repo-plugins/pull/1#issuecomment-373903664, but I can't see anything wrong in the log @TravisBuddy commented in both Mac and Linux: 

![image](https://user-images.githubusercontent.com/13808883/37867315-62da3f9a-2fa7-11e8-8ceb-425bc0197649.png)

@mzfr @Razzeee

**Comment by Razzeee** [comment]
I guess you need to grep for the color markup and just remove that. See here for the python code where we set it: https://github.com/xbmc/addon-check/blob/master/common.py#L2

**Comment by bluzi** [comment]
@Razzeee Could be, but look at the image I posted, I don't see the markup characters: 

![image](https://user-images.githubusercontent.com/13808883/37876046-ff6bb1b0-304f-11e8-93c5-29eec99d279d.png)

**Comment by Razzeee** [comment]
Yes, I would think that they get swallowed by the box element (maybe) along with those missing lines and letters.

**Comment by bluzi** [comment]
Don't think so, I still can't see it even though I removed the box:

![image](https://user-images.githubusercontent.com/13808883/37878360-b21ad7da-3070-11e8-8d7d-73e9aa74b2da.png)


We're removing non-ansi characters using `strip-ansi` here: https://github.com/bluzi/travis-buddy/blob/master/utils/utils.js#L28. Do you want to see those characters?

**Comment by bluzi** [comment]
@Razzeee, let's be on the same page: I thought you wanted those characters stripped. 
If you want I can make sure they'll appear, but there is no way to add colors to GitHub comments.

**Comment by Razzeee** [comment]
Yes, markup (as in color) characters stripped is fine. But for some reason we're loosing more characters along the way. Try reading the above screenshot, line 6: `Addon i icon exists`
In the original raw log that's two lines `Addon id matches foldername
Image icon exists`
Which somehow get mangled into that. Just as an example, the lines following that one get even crazier.

**Comment by bluzi** [comment]
Right, I see what you mean now. Will have a look soon.

**Comment by bluzi** [comment]
@Razzeee, [I've tried to reproduce the issue with a test](https://github.com/bluzi/name-db/pull/501#issuecomment-377710315), it doesn't reproduce. I'm getting the feeling that maybe this was a temporary issue with TravisCI's webhook payload. 

Can you please try and add another commit to https://github.com/Razzeee/repo-plugins/pull/1 and see if it still happens?

**Comment by Razzeee** [comment]
@bluzi seems to still be there

https://github.com/Razzeee/repo-plugins/pull/1#issuecomment-377777348
diff it with
https://travis-ci.org/Razzeee/repo-plugins/builds/360839581?utm_source=github_status&utm_medium=notification

**Comment by Razzeee** [comment]
Yay, looks like a complete log now
https://github.com/Razzeee/repo-plugins/pull/1#issuecomment-377940568

Can you share what was causing this?

**Comment by bluzi** [comment]
Glad to hear.

It's actually turned out to be a result of a weird behavior of TravisCI we're experiencing: Sometimes, TravisCI triggers the webhook before it finished to fully generate the build log. 
We've seen it before in a form of an incomplete build log that after few seconds turns into a complete log, and in your case it was a corrupted build log that after few seconds becomes a valid log. 

In other words: Waiting few seconds before requesting the log solved this.
I'll also make sure to open an issue in TravisCI's repository and tell them about this behavior.

**Comment by bluzi** [comment]
Created an issue: https://github.com/travis-ci/travis-ci/issues/9429

**Comment by bluzi** [comment]
@Razzeee @mzfr Do you guys need anything else in order to integrate? 
BTW, I advise you to take a look at https://github.com/bluzi/travis-buddy/blob/master/whats-new.md :)

**Comment by Razzeee** [comment]
So to have it print the same things on failure and success we would need two configs according to https://github.com/bluzi/travis-buddy/blob/master/whats-new.md#custom-templates ?
And what's the error state? When it fails on the travis side of things?

**Comment by bluzi** [comment]
Yup, you're going to need two templates. 
Error means something went wrong with the build itself, could be a TravisCI error, but could also be something wrong with the builds process that has nothing to do with the test script. 

https://github.com/travis-ci/travis-ci/issues/9376

**Comment by Razzeee** [comment]
Another question, just because it doesn't seem to be in the docs, will travisBuddy need repo access now to get the PR creators identity?

**Comment by bluzi** [comment]
Nope

**Comment by Razzeee** [comment]
Do successfull executions also set all variables? Doesn't seem like it?

https://github.com/xbmc/repo-plugins/pull/1783#issuecomment-380666802

**Comment by bluzi** [comment]
They don't set the log variables, I'll add support for it soon.
